pkg/sync/v6/data/PrometheusRule.monitoring.coreos.com/openshift-monitoring/prometheus-k8s-arorules.yaml

```diff
@@ -0,0 +1,86 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    prometheus: k8s
+    role: alert-rules
+  name: prometheus-k8s-arorules
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: aro-sync.rules
+    rules:
+    - alert: SyncPodDown
+      annotations:
+        description: Sync Pod is down
+        summary: Sync Pod is down
+      expr: |
+        absent(up{service="sync"} == 1)
+      for: 10m
+      labels:
+        severity: warning
+    - alert: SyncNotExecuted
+      annotations:
+        description: There was no Sync in the last 10min
+        summary: Sync not executed in last 10min
+      expr: |
+        time() - sync_last_executed > 600
+      labels:
+        severity: warning
+  - name: aro-canary.rules
+    rules:
+    - alert: CanaryPodDown
+      annotations:
+        description: At least one Canary Pod is down for 10min
+        summary: Canary Pod is down
+      expr: |
+        absent(sum(up{service="canary"}) == 2)
```

Comment:
I don't think `absent` is what we want to use here, at least if I'm reading the description right. If `sum(up{service="canary"})` = 1, for example, this won't alert. Actually, even if the `up` vector for the canary service doesn't exist at all, because we're calling `sum`, this will always return an empty vector, at least according to the [Prometheus docs](https://prometheus.io/docs/prometheus/latest/querying/functions/#absent): 

```
absent(sum(nonexistent{job="myjob"}))
# => {}
```

Hence, this will never alert. I think the query we actually want is

```
sum(up{service="canary"}) < 2
```

Reply:
@Makdaam I know that the out-of-the-box alerts that come with the installation use `absent` to detect services which are down. However I think its use together with `sum` is not correct as @ehashman points out. 

Personally, I would do the following 

```
avg without(instance)(up{job="canary"}) < 0.5
```
This will alert if more than half of the total number of canary instances are down. Such a query is resistant to changes in the actual topology of your service deployments i.e. it works if the canary job has 2 instances now or 100 instances in the future.

Reply:
@ehashman 
I tested the three scenarios with 3 alerts on a healthy cluster:
 - if the number of up metrics for a service is < expected number
 - if the number of up metrics for a service is > expected number
 - if there are no up metrics for a service

The expressions I used are:
absent(sum(up{service="canary"})  == 3)
absent(sum(up{service="canary"})  == 1)
absent(sum(up{service="heffalump"})  == 2)

All of them are active when there are 2 canary pods, and 0 heffalump pods/services.
The alert is active for every element in the returned vector https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/ 

The reason why I used absent is that sum(vector_expression) is an aggregation operator and not a function. If the vector_expression is an empty vector, sum will also be an empty vector, and not 0. Only when a vector_expression containing elements which sum up to 0 makes the sum 0. This means that if all pods are down and no up values are selected the alert doesn't trigger.

Reply:
@charlesakalugwu
I wanted to detect whenever the number of pods is different than expected (so the expression has to know how many pods are supposed to be up in a healthy cluster). 

It might be a good idea to add a higher severity alert when all canary pods are down.